src/mps/observer.jl

```diff
@@ -27,40 +27,48 @@ implements custom measurements and allows
 the `dmrg` function to return early if an
 energy convergence criterion is met.
 """
-struct DMRGObserver <: AbstractObserver
+struct DMRGObserver{T} <: AbstractObserver
   ops::Vector{String}
   sites::Vector{<:Index}
   measurements::Dict{String,DMRGMeasurement}
-  energies::Vector{Complex}
+  energies::Vector{T}
   truncerrs::Vector{Float64}
   etol::Float64
   minsweeps::Int64
 end
 
 """
     DMRGObserver(;energy_tol=0.0,
-                  minsweeps=2)
+                  minsweeps=2,
+                  complex_energies=false)
 
 Construct a DMRGObserver by providing the energy
 tolerance used for early stopping, and minimum number
 of sweeps that must be done.
 
+Optional keyword arguments:
   - energy_tol: if the energy from one sweep to the
     next no longer changes by more than this amount,
     stop after the current sweep
   - minsweeps: do at least this many sweeps
+  - complex_energies: allow stored energies to be 
+    complex for non-Hermitian DMRG
 """
-function DMRGObserver(; energy_tol=0.0, minsweeps=2)
+function DMRGObserver(; energy_tol=0.0, minsweeps=2, complex_energies=false)
```

Comment:
I'm now seeing the issue that a user will define the `DMRGObserver` before DMRG starts and we don't want to force them to tell us if the energy will be complex beforehand.

An alternative could be:
```julia
function DMRGObserver(; energy_tol=0.0, minsweeps=2, energy_type=Number)
  return DMRGObserver([], Index[], Dict{String,DMRGMeasurement}(), energy_type[], Float64[], energy_tol, minsweeps)
end
```
Then a user can specify the type as more specialized if they want (and additionally that would cover cases like `Float32`).

It may just be simpler to go back to your original design of using `Vector{Number}` and not parameterizing the type, my main concern there is that analyzing the energies might be slow for `Vector{Number}` but someone could always convert to `Vector{Float64}` as needed.

In [Observers.jl](https://github.com/GTorlai/Observers.jl) we keep the results in a concretely-typed vector (if possible) but don't parametrize the `Observer` type by the results. In that way we can update the storage type dynamically (kind of like the ITensor storage type):
```julia
julia> using Observers

julia> Measure1(; x) = x;

julia> Measure2(; y) = y;

julia> obs = Observer([Measure1, Measure2]);

julia> results(obs, "Measure1")
Any[]

julia> results(obs, "Measure2")
Any[]

julia> update!(obs; x=1.1, y=1.2);

julia> results(obs, "Measure1")
1-element Vector{Float64}:
 1.1

julia> results(obs, "Measure2")
1-element Vector{Float64}:
 1.2

julia> update!(obs; x=2.1, y=2.2 + 2.3im);

julia> results(obs, "Measure1")
2-element Vector{Float64}:
 1.1
 2.1

julia> results(obs, "Measure2")
2-element Vector{ComplexF64}:
 1.2 + 0.0im
 2.2 + 2.3im
```
Anyway, that kind of solution is overkill here and I don't think it is worth getting too fancy with the `DMRGObserver` type since it should get superseded by `Observers.jl` eventually.

Reply:
Great. In the end your idea wasn't hard to implement using a keyword argument, and I like `energy_type` better than what I did because the code is both cleaner and it's more extensible to other cases. But yes this is all probably a holdover for now until we introduce Observers.jl soon anyway.